fix: round after math

Westbrook opened the pull request:
Fixes #1300
Fixes #1410
Refs #1400

Rounding was previously happening to _ONE_ of the measured inputs and then other inputs were not, so the output wouldn't actually be rounded. This moves the rounding to _only_ happen at the end of the style calculations so that it ensure the styles sent the the client are rounded.

atomiks commented:
Does it also help this https://github.com/popperjs/popper-core/issues/1410?

Westbrook commented:
It definitely looks a lot better. "Solved" is tricky when _somethings_ can be subpixel and some things can't be. While zooming in continues to throw a big fat wrench, non-zoomed this appears to align things as expect in the case of #1410 as well.

atomiks commented:
> While zooming in continues to throw a big fat wrench

What about in combo with the DPR PR too https://github.com/popperjs/popper-core/pull/1400

Westbrook commented:
DPR doesn't seem to change while zooming. I'm not sure there is any way to surface zoom in the client to account for it either. I'll look into that.

Westbrook commented:
My mistake, `window.devicePixelRatio` does seem to follow the zoom, the issue is when during the 110% zoom where the math still ends up with a non-representable pixel fraction. However, even in that case it looks much better than what was being previously delivered. All other zooms seem to be good to go in the provided demo.

Westbrook commented:
Any thoughts on this test failure? I’m not sure how it was considered passing before based on the code I’m seeing. Happy to update the baseline with some agreement there. This update helps my project a lot and I’d love for it to come without having to patch the package.

atomiks commented:
Regarding that test, a difference of 4px seems kinda odd (I'd expect 1px max), do you have a replication of the differences?

FezVrasta commented:
The test asserts that 100px width + 10.3px offset + 2px should result into 112px after it gets rounded. 108 seems not to be correct, it looks like you are subtracting 2 rather than adding them?

Westbrook commented:
I stepped through this some more and better understand what is going on, but I'm not sure what to say is the right answer.

The `roundOffsets` method here is:

```
roundOffsets: ({ x, y }) => ({
        x: Math.round(x + 2),
        y: Math.round(y + 2),
}),
```

If you round at the beginning, you'll end up with `-112` because you add 2 to the input (`10.3` to `12`), and if you round at the end you get `-108` because you round the output (`-110.3` to `-108`).

What is the _meaning_ of this test? `// customize roundOffsets impls` doesn't really explain why you are making a custom `roundOffset` to tell what is _supposed_ to be happening here. However, it does imply that my alteration would be a breaking change for some, regardless of it being more or less correct output. Rounding at the beginning and at the end would be less accurate, but would allow for this to both function as before and ensure all number sent to the client are rounded correctly. OR, we could allow for all custom `roundOffset` code to be run _only_ at the beginning, as that likely means the user is getting what they'd prefer here already?

Westbrook commented:
Or maybe use the `roundOffsets` option at the beginning of the process as available, but not as a replacement for `roundOffsetsByDPR`, which should likely be renamed to `roundPlacementByDPR` and run 100% of the time at the end of `mapToStyles`?

Westbrook commented:
Split the two parts of rounding outlined in https://github.com/popperjs/popper-core/pull/1418#issuecomment-981094250 to before and after the computations and added a new test (leveraging measurements from https://stackblitz.com/edit/react-nyp2yj?file=src%2FApp.js) that failed to deliver DPR correct computations when using the previous method of rounding (which is the basis for #1410).

Westbrook commented:
Any insight on the Flow issue would be most helpful. That same command doesn't fail locally and I've never used Flow outside of this project.

Westbrook commented:
Oh, I think I got it now. My branch was out of date 😅 Hopefully this is the one!

FezVrasta commented:
Thanks! Merging.

Westbrook commented:
Can we get a release on this to hold people over until they're able to update to Floating UI?
